### PR TITLE
Default resource limits for build pods

### PR DESCRIPTION
**Pull request by wanghaoran1988**
[Trello Card](https://trello.com/c/fdRnlbqn/1035-default-resource-limits-for-build-pods)

### CODE REVIEW

**Comment by wanghaoran1988** [comment]
@bparees Please help review, when you can ,thanks.

**Comment by bparees** [comment]
@wanghaoran1988 I appreciate the initiative but please start consulting with me before starting PRs for backlog features, especially ones that haven't even been groomed (ie are still in the new list).  I don't want you to waste your time on something we aren't sure we want, or haven't thought through the direction on yet.

This one is ok but it's going to cause merge conflicts with this PR https://github.com/openshift/origin/pull/11144 which is for a feature that's committed for 3.4, so I don't want to merge your PR until that goes through.

File: pkg/build/admission/defaults/admission_test.go [hunk 1]
```diff
@@ -130,3 +130,143 @@ func TestIncrementalDefaults(t *testing.T) {
...
+	}{
+		"BuildDefaults plugin and BUILD object both defined resource limits and requests": {
+			DefaultResource: kapi.ResourceRequirements{
+				Limits: kapi.ResourceList{
+					kapi.ResourceName(kapi.ResourceCPU):    resource.MustParse("10"),
+					kapi.ResourceName(kapi.ResourceMemory): resource.MustParse("1G"),
+				},
+				Requests: kapi.ResourceList{
+					kapi.ResourceName(kapi.ResourceCPU):    resource.MustParse("10"),
+					kapi.ResourceName(kapi.ResourceMemory): resource.MustParse("1G"),
```
Comment (bparees) [inline pkg/build/admission/defaults/admission_test.go]: for this and all the other tests, use different values for "requests" and "limits" to ensure the values aren't getting mixed up.  As it is if you were assigning the limit value to the resource field, the test would still pass.

File: pkg/build/admission/defaults/admission_test.go [hunk 2]
```diff
@@ -130,3 +130,143 @@ func TestIncrementalDefaults(t *testing.T) {
...
+					kapi.ResourceName(kapi.ResourceCPU):    resource.MustParse("20"),
+					kapi.ResourceName(kapi.ResourceMemory): resource.MustParse("2G"),
+				},
+				Requests: kapi.ResourceList{
+					kapi.ResourceName(kapi.ResourceCPU):    resource.MustParse("10"),
+					kapi.ResourceName(kapi.ResourceMemory): resource.MustParse("1G"),
+				},
+			},
+		},
+		"BuildDefaults plugin defined nothings ,BUILD object defined resource limits": {
```
Comment (bparees) [inline pkg/build/admission/defaults/admission_test.go]: "BuildDefaults plugin defined nothing, Build object defined resource limits"

(nothing is singular, comma goes immediately after nothing, Build doesn't need to be all caps.  Same changes in the other test descriptions)

File: pkg/build/admission/defaults/admission_test.go [hunk 3]
```diff
@@ -130,3 +130,143 @@ func TestIncrementalDefaults(t *testing.T) {
...
+			ExpectedResource: kapi.ResourceRequirements{
+				Limits: kapi.ResourceList{
+					kapi.ResourceName(kapi.ResourceCPU):    resource.MustParse("20"),
+					kapi.ResourceName(kapi.ResourceMemory): resource.MustParse("2G"),
+				},
+				Requests: kapi.ResourceList{
+					kapi.ResourceName(kapi.ResourceCPU):    resource.MustParse("20"),
+					kapi.ResourceName(kapi.ResourceMemory): resource.MustParse("2G"),
+				},
+			},
```
Comment (bparees) [inline pkg/build/admission/defaults/admission_test.go]: Needs a test where the build Resources object is empty to test your length checking/list initializing logic in the defaulter.

File: pkg/build/admission/defaults/api/types.go [hunk 4]
```diff
@@ -25,6 +25,9 @@ type BuildDefaultsConfig struct {
 	// SourceStrategyDefaults are default values that apply to builds using the
 	// source strategy.
 	SourceStrategyDefaults *SourceStrategyDefaultsConfig
+
+	// Resources computes resource requirements to execute the build.
+	Resources kapi.ResourceRequirements
```
Comment (bparees) [inline pkg/build/admission/defaults/api/types.go]: should be a pointer

Comment (bparees) [inline pkg/build/admission/defaults/api/types.go]: and then you will need to add a test where Resources is nil (in addition to the existing test where Resources is non-nil but empty)

**Comment by wanghaoran1988** [comment]
@bparees I am very sorry to work on the not groomed card without consulting you, and upset your plan, I promise this will never happen again, and sure , you can block this one  until [#11144](https://github.com/openshift/origin/pull/11144) is merged.

**Comment by bparees** [comment]
@wanghaoran1988 this can be resumed now that the 3.4 work has landed, but i'm afraid you're going to have to do a bit of a messy rebase first off.

**Comment by wanghaoran1988** [comment]
use this[11575](https://github.com/openshift/origin/pull/11575) instead

**Pull request closed without merging**